### PR TITLE
RLP Encoding / Decoding

### PR DESCRIPTION
**Pull request by bbondy**
RLP is the main encoding method used to serialize objects in Ethereum. We need it for dealing with transactions and more.

Resolves https://github.com/brave/brave-browser/issues/15049

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/brave_wallet_utils.cc [hunk 1]
```diff
@@ -189,4 +189,203 @@ std::unique_ptr<std::vector<uint8_t>> MnemonicToSeed(
   return rv == 1 ? std::move(seed) : nullptr;
 }
 
+bool RLPToInteger(const std::string& s, size_t* val) {
```
Comment (darkdh) [inline components/brave_wallet/brave_wallet_utils.cc:192]: maybe put these internal functions into anonymous namespace.

File: components/brave_wallet/browser/rlp_decode.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+#define BRAVE_COMPONENTS_BRAVE_WALLET_RLP_DECODE_H_
+
+#include <string>
+
+#include "base/values.h"
+
+namespace brave_wallet {
+
+// Recursive Length Prefix (RLP) decoding of arbitrarily nested arrays of data
+bool RLPDecode(const std::string& s, base::Value* output);
```
Comment (darkdh) [inline components/brave_wallet/browser/rlp_decode.h:16]: I can see 0x is stripped in test cases so it is better to specify the input format in comment.

File: components/brave_wallet/browser/rlp_decode_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+TEST(RLPDecodeTest, ComplexStructure) {
+  base::Value val;
+  ASSERT_TRUE(RLPDecode(FromHex("0xe383636174ca85707570707983636f7785686f727365"
+                                "c1c083706967c180857368656570"),
+                        &val));
+  ASSERT_EQ("['cat', ['puppy', 'cow'], 'horse', [[]], 'pig', [''], 'sheep']",
+            RLPTestValueToString(val));
+}
+
+}  // namespace brave_wallet
```
Comment (darkdh) [inline components/brave_wallet/browser/rlp_decode_unittest.cc:279]: we should also add some invalid inputs like and expect the result to be false and the `base::Value::is_none` to be true
https://github.com/ethereum/tests/blob/dd23a3e4edd4619d99933fc135cc88b31ca3cafb/RLPTests/invalidRLPTest.json

**Review by darkdh — COMMENTED** [review]
good work

File: components/brave_wallet/browser/rlp_decode.cc [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+bool RLPDecode(const std::string& s, base::Value* output) {
+  size_t offset;
+  size_t data_len;
+  return RLPDecodeInternal(s, output, &offset, &data_len);
```
[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/browser/rlp_decode.cc:159]: it's for recursive call of `RLPDecodeInternal`

File: components/brave_wallet/browser/rlp_decode.cc [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+  return true;
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+bool RLPDecode(const std::string& s, base::Value* output) {
+  size_t offset;
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/brave_wallet/browser/rlp_decode.cc:152]: Fixed 51302353876330564318ed3f22ff8bf7bc1f99e9

File: components/brave_wallet/browser/rlp_decode.h (hunk 2 above)
Comment (bbondy) [reply components/brave_wallet/browser/rlp_decode.h:17]: Fixed in 51302353876330564318ed3f22ff8bf7bc1f99e9

File: components/brave_wallet/browser/rlp_encode.cc [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    base::Value::BlobStorage blob = val.GetBlob();
+    std::string s(blob.begin(), blob.end());
+    if (blob.size() == 1 && static_cast<uint8_t>(blob[0]) < 0x80) {
+      return s;
+    }
+    return RLPEncodeLength(blob.size(), 0x80) + s;
+  } else if (val.is_string()) {
+    std::string s;
+    val.GetAsString(&s);
+    base::Value::BlobStorage blob(s.begin(), s.end());
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/brave_wallet/browser/rlp_encode.cc]: Fixed in 51302353876330564318ed3f22ff8bf7bc1f99e9

File: components/brave_wallet/browser/rlp_decode.cc [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+  size_t i;
+  if (1 + prefix - 0xb7 <= length && RLPToInteger(s.substr(1, prefix - 0xb7), &i)) {
+    if (prefix <= 0xbf && length > prefix - 0xb7 &&
+        length > prefix - 0xb7 + i) {
+      size_t lenOfStrLen = prefix - 0xb7;
+      size_t strLen;
+      if (1 + lenOfStrLen > length || !RLPToInteger(s.substr(1, lenOfStrLen), &strLen)) {
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/brave_wallet/browser/rlp_decode.cc]: Fixed in 51302353876330564318ed3f22ff8bf7bc1f99e9

File: components/brave_wallet/browser/rlp_decode.cc [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    *offset = 1;
+    *data_len = listLen;
+    *value = base::ListValue();
+    return true;
+  }
+
+  if (prefix - 0xf7 < 1 ||
+      1 + prefix - 0xf7 > length || !RLPToInteger(s.substr(1, prefix - 0xf7), &i)) {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/brave_wallet/browser/rlp_decode.cc]: Fixed in 51302353876330564318ed3f22ff8bf7bc1f99e9

File: components/brave_wallet/browser/rlp_decode.cc [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  if (prefix <= 0xff && length >= prefix - 0xf7 && length >= prefix - 0xf7 + i) {
+    size_t lenOfListLen = prefix - 0xf7;
+    size_t listLen;
+    if (1 + lenOfListLen > length || !RLPToInteger(s.substr(1, lenOfListLen), &listLen)) {
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/brave_wallet/browser/rlp_decode.cc]: Fixed in 51302353876330564318ed3f22ff8bf7bc1f99e9

File: components/brave_wallet/browser/rlp_decode_unittest.cc (hunk 3 above)
Comment (darkdh) [reply components/brave_wallet/browser/rlp_decode_unittest.cc:279]: Also we can add a case with nullptr of base::Value and expect the result to be false

File: components/brave_wallet/browser/rlp_encode_unittest.cc [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+
+TEST(RLPEncodeTest, ComplexStructure) {
+  std::string v = brave_wallet::RLPEncode(RLPTestStringToValue(
+      "['cat',['puppy', 'cow'], 'horse', [[]], 'pig', [''], 'sheep']"));
+  ASSERT_EQ(ToHex(v),
+            "0xe383636174ca85707570707983636f7785686f727365c1c083706967c1808573"
+            "68656570");
+}
+
+}  // namespace brave_wallet
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/brave_wallet/browser/rlp_encode_unittest.cc:393]: Fixed in e6d6a9667fbb488969fadeba06c457c745d31c5c

File: components/brave_wallet/browser/rlp_encode_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+namespace {
+
+base::Value RLPTestStringToValue(const std::string& s, std::string* remaining) {
+  const char* start = s.c_str();
+  const char* p = start;
+  enum State { Start, InString, InNumber, InList };
+  size_t list_depth = 0;
+  State state = Start;
+  while (*p != '\0') {
+    if (*p >= '0' && *p <= '9' && state == Start) {
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/brave_wallet/browser/rlp_encode_unittest.cc]: Fixed in e6d6a9667fbb488969fadeba06c457c745d31c5c

File: components/brave_wallet/browser/rlp_decode.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_RLP_DECODE_H_
```
Comment (SergeyZhukovsky) [inline components/brave_wallet/browser/rlp_decode.h:6]: has to be `BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_RLP_DECODE_H_`

File: components/brave_wallet/browser/rlp_encode.cc [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+base::Value RLPUint256ToBlobValue(uint256_t input) {
+  base::Value::BlobStorage output;
+  while (input > static_cast<uint256_t>(0)) {
+    uint8_t i = static_cast<uint8_t>(input & static_cast<uint256_t>(0xFF));
+    output.push_back(static_cast<char>(i));
+    input >>= 8;
+  }
+  // Return the vector reversed (big endian)
+  std::reverse(output.begin(), output.end());
```
Comment (SergeyZhukovsky) [inline components/brave_wallet/browser/rlp_encode.cc:44]: algorithm header

File: components/brave_wallet/browser/rlp_encode.h [hunk 14]
```diff
@@ -0,0 +1,25 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_RLP_ENCODE_H_
```
Comment (SergeyZhukovsky) [inline components/brave_wallet/browser/rlp_encode.h:6]: invalid ifndef name

File: components/brave_wallet/browser/rlp_decode.cc [hunk 15]
```diff
@@ -85,17 +81,13 @@ bool RLPDecodeLength(const std::string& s,
...
-    size_t lenOfListLen = prefix - 0xf7;
-    size_t listLen;
-    if (1 + lenOfListLen > length || !RLPToInteger(s.substr(1, lenOfListLen), &listLen)) {
-      return false;
-    }
-    *offset = 1 + lenOfListLen;
+  if (prefix - 0xf7 >= 1 &&
+      prefix <= 0xff &&
+      length >= 1 + prefix - 0xf7 &&
+      length >= prefix - 0xf7 + 1 &&
```
Comment (darkdh) [inline components/brave_wallet/browser/rlp_decode.cc:87]: These are equivalent

File: components/brave_wallet/browser/rlp_decode.cc [hunk 16]
```diff
@@ -85,17 +81,13 @@ bool RLPDecodeLength(const std::string& s,
...
-    return false;
-  }
-  if (prefix <= 0xff && length >= prefix - 0xf7 && length >= prefix - 0xf7 + i) {
-    size_t lenOfListLen = prefix - 0xf7;
-    size_t listLen;
-    if (1 + lenOfListLen > length || !RLPToInteger(s.substr(1, lenOfListLen), &listLen)) {
-      return false;
-    }
-    *offset = 1 + lenOfListLen;
+  if (prefix - 0xf7 >= 1 &&
```
Comment (darkdh) [inline components/brave_wallet/browser/rlp_decode.cc:84]: This is not necessary,  previous if condition would handle this

File: components/brave_wallet/browser/rlp_encode.h (hunk 14 above)
Comment (bbondy) [reply components/brave_wallet/browser/rlp_encode.h:6]: Fixed by interactive rebase in the next push

File: components/brave_wallet/browser/rlp_encode.cc (hunk 13 above)
Comment (bbondy) [reply components/brave_wallet/browser/rlp_encode.cc:44]: Fixed by interactive rebase in the next push

File: components/brave_wallet/browser/rlp_decode.h (hunk 12 above)
Comment (bbondy) [reply components/brave_wallet/browser/rlp_decode.h:6]: Fixed by interactive rebase in the next push

File: components/brave_wallet/browser/rlp_decode.cc (hunk 15 above)
Comment (bbondy) [reply components/brave_wallet/browser/rlp_decode.cc:87]: Fixed by interactive rebase in the next push

**Review by darkdh — APPROVED** [review]
++

**Pull request merged**